src/main.c

```diff
@@ -1,69 +1,123 @@
 #include <stdlib.h>
 #include <stdio.h>
 
-struct node {
-    struct node *pnext;
-    int item;
-};
+/* node data is stored here */
+#include "node.h"
 
-/*
 void setItem(struct node *node, int item) {
-    node -> item = item;
+    node->item = item;
 }
-*/
 
-/*
-void addNext(struct node *current, struct node next) {
+void addNext(struct node *current, struct node *next) {
     if(current -> pnext != NULL) {
-        current -> pnext = &next;
+        /**
+         * Simply taking a `struct node` and adding it with the dereference
+         * operator would not work here!
+         * This is because when you pass a struct node, it is passed by value.
+         * i.e. `next` would live in this functions stack.
+         *
+         * Therefore, when you assign `current->pnext = &next` it references
+         * this functions _own stack_ which will go away after the function
+         * returns.
+         *
+         * Therefore this function must take a `struct node *next` where the
+         * address is given, and set it to that.
+         */
+        current->pnext = next;
     } else {
-        //This doesn't behave how I expected it to
-        #warning attempting to add 'next' to a node that already has next
+        /**
+         * Preprocessor warnings occur whenever the preprocessor sees them. This
+         * occurs before compilation, and has nothing to do with actual source
+         * code. This is why preprocessor declarations should always be put at
+         * the beginning of a line.
+         * #warning creates a compilation warning—not a runttime one
+         */
+        fprintf(stderr, "error: "
+                "attempting to add 'next' to a node that already has next\n");
     }
 }
-*/
 
-int main() {
-    //TODO get data from user
-    int data[] = {1, 2, 3, 4, 5, 6, 7, 8, 9}, i = 0;
-    //will this work
-    struct node *head = malloc(sizeof(struct node)), *current = head;
-    //or do we need this
-    //struct node *head = (struct node *) malloc(sizeof(struct node));
+/**
+ * Call function main as follows: ./main 1 2 3 4 5 6 7 8...
+ */
+int main(int argc, char **argv) {
+    if (argc == 1) {
+        fprintf(stderr, "Need at least 1 argument!", argc);
+        exit(EXIT_FAILURE);
+    }
+
+    fprintf(stderr, "main called with %d arguments\n", argc);
+    /**
+     * Create an array to hold the arguments (not really necessary, but for
+     * demonstration how to use calloc.
+     * calloc is better than malloc since it automatically initializes its data
+     * to `0`. with malloc the data is uninitialized!
+     */
+    int length = argc-1;
+    int *data = calloc(sizeof(int), argc);
+    /* Get user input. note: first argument is at argv[1] */
```

Comment:
why is first argument at `argv[1]`?

Reply:
This is a weird part of posix. argv[0] is always the name of the program

think about when you run programs from the command line, the first part of the command is always the program name.
so argv is pretty much if you take each space-separated list of strings in the command you ran and put them all in an array